modules/ray_tracing/src/userobjects/RayTracingStudy.C

```diff
@@ -0,0 +1,1794 @@
+    ids[i] = registerRay(names[i], graceful);
+  return ids;
+}
+
+RayID
+RayTracingStudy::registeredRayID(const std::string & name, const bool graceful /* = false */) const
+{
+  Threads::spin_mutex::scoped_lock lock(_spin_mutex);
+
+  if (!_use_ray_registration)
+    mooseError(_error_prefix, ": Should not use registeredRayID() with Ray registration diabled");
+
+  const auto search = _registered_ray_map.find(name);
+  if (search != _registered_ray_map.end())
+    return search->second;
+
+  if (graceful)
+    return Ray::INVALID_RAY_ID;
+
+  mooseError(_error_prefix,
+             ": Attempted to obtain ID of registered Ray ",
+             name,
+             ", but a Ray with said name is not registered.");
+}
+
+const std::string &
+RayTracingStudy::registeredRayName(const RayID ray_id) const
+{
+  Threads::spin_mutex::scoped_lock lock(_spin_mutex);
+
+  if (!_use_ray_registration)
+    mooseError(_error_prefix, ": Should not use registeredRayName() with Ray registration diabled");
+
+  const auto search = _reverse_registered_ray_map.find(ray_id);
+  if (search != _reverse_registered_ray_map.end())
+    return search->second;
+
+  mooseError(_error_prefix,
+             ": Attempted to obtain name of registered Ray with ID ",
+             ray_id,
+             ", but a Ray with said ID is not registered.");
+}
+
+Real
+RayTracingStudy::computeTotalVolume()
+{
+  Real volume = 0;
+  for (const auto & elem : *_mesh.getActiveLocalElementRange())
+    volume += elem->volume();
+  _communicator.sum(volume);
+  return volume;
+}
+
+const std::vector<std::vector<BoundaryID>> &
+RayTracingStudy::getInternalSidesets(const Elem * elem) const
+{
+  mooseAssert(_use_internal_sidesets, "Not using internal sidesets");
+  mooseAssert(hasInternalSidesets(), "Processor does not have internal sidesets");
+  mooseAssert(_internal_sidesets_map.size() > _elem_index_helper.getIndex(elem),
+              "Internal sideset map not initialized");
+
+  const auto index = _elem_index_helper.getIndex(elem);
+  return _internal_sidesets_map[index];
+}
+
+TraceData &
+RayTracingStudy::initThreadedCachedTrace(const std::shared_ptr<Ray> & ray, THREAD_ID tid)
+{
+  mooseAssert(shouldCacheTrace(ray), "Not caching trace");
+  mooseAssert(currentlyPropagating(), "Should only use while tracing");
+
+  _threaded_cached_traces[tid].emplace_back(ray);
+  return _threaded_cached_traces[tid].back();
+}
+
+void
+RayTracingStudy::verifyRay(const std::shared_ptr<Ray> & ray,
+                           const THREAD_ID tid,
+                           const std::string & error_prefix)
+{
+  if (!ray)
+    mooseError(error_prefix, ": Cannot verify null Ray");
+
+  const auto error = [this, &error_prefix, &ray](const std::string & error) {
+    mooseError(error_prefix, ":\n\nRay ", error, "\n\n", ray->getInfo(this));
+  };
+
+  if (ray->hasTraced())
+    error("has uncleared counters");
+
+  if (ray->invalidID())
+    error("has an invalid ID");
+
+  if (ray->invalidStartPoint())
```

Comment:
I think a lot of stuff in this verify function seems like it should be asserts.  For example, the only way your start could be invalid based on the code that I looked at is if it was never set at all - which means the developer wrote broken code.